python/pyiceberg/table/__init__.py

```diff
@@ -934,206 +963,611 @@ def case_sensitive(self, case_sensitive: bool) -> UpdateSchema:
-        if element_type is None:
-            raise ValueError(f"Cannot delete element type from list: {element_field}")
+        # handle add & moves
+        added = self._adds.get(field.field_id)
+        moves = self._moves.get(field.field_id)
+        if added is not None or moves is not None:
+            if not isinstance(field.field_type, StructType):
+                raise ValueError(f"Cannot add fields to non-struct: {field}")
 
-        is_element_optional = not list_type.element_required
+            if new_fields := _add_and_move_fields(field.field_type.fields, added or [], moves or []):
+                return StructType(*new_fields)
 
-        if is_element_optional == element_field.required and list_type.element_type == element_type:
-            return list_type
+        return field_result
+
+    def list(self, list_type: ListType, element_result: Optional[IcebergType]) -> Optional[IcebergType]:
+        element_type = self.field(list_type.element_field, element_result)
+        if element_type is None:
+            raise ValueError(f"Cannot delete element type from list: {element_result}")
 
-        return ListType(list_type.element_id, element_type, is_element_optional)
+        return ListType(element_id=list_type.element_id, element=element_type, element_required=list_type.element_required)
 
-    def map(self, map_type: MapType, key_result: IcebergType, value_result: IcebergType) -> IcebergType:
+    def map(
+        self, map_type: MapType, key_result: Optional[IcebergType], value_result: Optional[IcebergType]
+    ) -> Optional[IcebergType]:
         key_id: int = map_type.key_field.field_id
-        if key_id in self.adds:
+        if key_id in self._adds:
             raise ValueError(f"Cannot add fields to map keys: {map_type}")
 
         value_field: NestedField = map_type.value_field
         value_type = self.field(value_field, value_result)
         if value_type is None:
             raise ValueError(f"Cannot delete value type from map: {value_field}")
 
-        is_value_optional = not map_type.value_required
+        return MapType(
+            key_id=map_type.key_id,
+            key_type=map_type.key_type,
+            value_id=map_type.value_id,
+            value_type=value_type,
+            value_required=map_type.value_required,
+        )
 
-        if is_value_optional != value_field.required and map_type.value_type == value_type:
-            return map_type
+    def primitive(self, primitive: PrimitiveType) -> Optional[IcebergType]:
+        return primitive
 
-        return MapType(map_type.key_id, map_type.key_field, map_type.value_id, value_type, not is_value_optional)
 
-    def primitive(self, primitive: PrimitiveType) -> IcebergType:
-        return primitive
+def _add_fields(fields: Tuple[NestedField, ...], adds: Optional[List[NestedField]]) -> Optional[Tuple[NestedField, ...]]:
+    adds = adds or []
+    return None if len(adds) == 0 else fields + tuple(adds)
```

Comment:
The way this is called, do we want to return None? `_add_and_move_fields` doesn't check for it.

Reply:
No, I think we can drop the `None`